### PR TITLE
Paid to paid upgrades

### PR DESCRIPTION
**Pull request by davidrapson**
This PR is a combination of work from @jennysivapalan and @wpf500 enabling paid-to-paid upgrades and updates to the front-end to make this work. Front-end updates based on wires from @davidmcdowell 

**New paid-to-paid upgrade screen**

upgrade to patron

**Upgrade thank you**

upgrade to patron - thank you

### CODE REVIEW

File: frontend/app/views/fragments/user/billingAddress.scala.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+@(userFields: model.PrivateFields)
+
+@if(userFields.billingAddress1) {
+    @fragments.user.deliveryAddress(userFields)
+} else {
+    <div class="address">
```
Comment (mattandrews) [inline frontend/app/views/fragments/user/billingAddress.scala.html]: `<address>` :sob:

Comment (davidrapson) [inline frontend/app/views/fragments/user/billingAddress.scala.html]: Hah, good point.

File: frontend/app/views/tier/upgrade/paidToPaid.scala.html [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+@import org.joda.time.DateTime
+@import views.support.Prices._
+@import views.support.Dates._
+@import views.support.Asset
+
+@main("Upgrade to " + targetTier, pageInfo=pageInfo) {
+
+    <main role="main" class="page-content l-constrained">
+
+        @fragments.page.pageHeader("Great, glad to see you've decided to become a " + targetTier)
```
Comment (mattandrews) [inline frontend/app/views/tier/upgrade/paidToPaid.scala.html]: This is just a personal thing so ignore the following sentence, but I *hate* this copy (it's in a few other places), specifically the "glad to see" bit. It implies a speaker ("*I'm* glad to see...") and it's not clear *who* is glad to see you're joining. I feel better now for having this moan. Thanks for coming on this journey with me. Now carry on.

File: frontend/app/views/tier/upgrade/paidToPaid.scala.html [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+        <div class="page-section page-section--bordered">
+            <div class="page-section__lead-in">
+                <h2 class="page-section__headline">What happens now</h2>
+            </div>
+            <div class="page-section__content">
+                <p>When you upgrade we want to make sure you are charged the right amount. We will charge for your <strong>@currentTier</strong> membership up until today and the amount remaining will be deducted from you're first payment as a <strong>@targetTier</strong>.</p>
```
Comment (mattandrews) [inline frontend/app/views/tier/upgrade/paidToPaid.scala.html]: Typo here: "you're" should be "your". Tut tut, copywriters.

File: frontend/app/views/tier/upgrade/paidToPaid.scala.html [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                <p><strong>@targetTier</strong> payments start today and will re-occur @{currentSubscription.paymentPeriodLabel}ly.</p>
```
Comment (mattandrews) [inline frontend/app/views/tier/upgrade/paidToPaid.scala.html]: More language pedantry but "re-occur" should be "recur" (cf. [Grammarist](http://grammarist.com/usage/recur-reoccur/)).

File: frontend/app/views/tier/upgrade/paidToPaid.scala.html (hunk 2 above)
Comment (davidrapson) [inline frontend/app/views/tier/upgrade/paidToPaid.scala.html]: Fairly sure it's this Who that's glad to see you:

tumblr_inline_n7ug88zvg11qegzff

**Comment by mattandrews** [comment]
:+1: besides spelling and grammar.

**Comment by jennysivapalan** [comment]
@jamesoram this needs a run through the tests and we probably want a functional test for paid to paid upgrade?
@rtyley would you be able to look over the Scala code?

**Comment by davidrapson** [comment]
@jamesoram Give me a shout if you need some hooks on the new template for your tests

File: frontend/app/controllers/TierController.scala [hunk 5]
```diff
@@ -43,19 +42,30 @@ trait UpgradeTier {
 
   def upgrade(tier: Tier) = MemberAction.async { implicit request =>
 
-    def futureCustomerOpt = request.member match {
+    def futurePaidPreviewOpt = request.member match {
       case paidMember: PaidMember =>
-        request.touchpointBackend.stripeService.Customer.read(paidMember.stripeCustomerId).map(Some(_))
+        for {
+          preview <- MemberService.previewUpgradeSubscription(paidMember, request.user, tier)
+          customer <- request.touchpointBackend.stripeService.Customer.read(paidMember.stripeCustomerId)
```
Comment (rtyley) [inline frontend/app/controllers/TierController.scala]: @jennysivapalan exciting (or sad) fact about Scala future for-comprehensions is these two futures - `previewUpgradeSubscription` & `Customer.read` - will not run concurrently as written here. See the 'For comprehensions' section in this document http://danielwestheide.com/blog/2013/01/09/the-neophytes-guide-to-scala-part-8-welcome-to-the-future.html

File: frontend/app/controllers/TierController.scala [hunk 6]
```diff
@@ -43,19 +42,30 @@ trait UpgradeTier {
...
     }
 
+    val subscriptionService = request.touchpointBackend.subscriptionService
+
     if (request.member.tier < tier) {
       for {
-        customerOpt <- futureCustomerOpt
+        paidPreviewOpt <- futurePaidPreviewOpt
+        subscriptionStatus <- subscriptionService.getSubscriptionStatus(request.member)
+        currentSubscription <- subscriptionService.getSubscriptionDetails(subscriptionStatus.current)
```
Comment (rtyley) [inline frontend/app/controllers/TierController.scala]: @jennysivapalan same problem with Scala future for-comprehensions is going on here - they're executing in series

File: frontend/app/controllers/TierController.scala (hunk 5 above)
Comment (jennysivapalan) [inline frontend/app/controllers/TierController.scala]: oooh interesting. I'll change this to instantiate futures before the for comprehension.

**Comment by jamesoram** [comment]
I will add a functional test for paid to paid upgrades.

File: frontend/app/controllers/TierController.scala (hunk 5 above)
Comment (jennysivapalan) [inline frontend/app/controllers/TierController.scala]: Done

**Comment by jamesoram** [comment]
it looks good so far, I just need to wait until tomorrow to confirm that it behaves correctly with users who have signed up on a different day.

**Comment by jamesoram** [comment]
By the way after the upgrade the top left corner (the tier under the username) does not get updated

**Comment by jamesoram** [comment]
Zuora takes some time to update the tier (this will be chased by Jason)

**Comment by jamesoram** [comment]
:+1:

**Pull request merged**